Delete loqusdb case and reset info in statusdb

fevac opened the pull request:
## Description

Related to https://github.com/Clinical-Genomics/cg/issues/1311

Currently, when deleting a case in loqusDB the corresponding entry in StatusDB is not removed automatically and you have to do it manually from cg-statusdb GUI. We want to be able to delete a case from loqusDB via cg and reset automatically statusDB.

### Added

- `delete observations` option to remove a case from LoqusDB: variants related to a case_id are removed from LoqusDB and the loqusdb_id in statusDB is reset to None 

### Changed

- `reset` option has been removed and replaced by `delete observations` option (see above).


### How to prepare for test

- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh delete-loqusdb-case`

#### Upload a case to loqusdb-stage. 
- [x] `$ cg upload observations --case_id <case_id>`
- [x] On statusdb-stage, under the tab `Sample`, check that `Loqusdb Id` is not empty
- [x] On hasta-stage, check that variants have been added to loqusdb `$ loqusdb cases --case-id <case_id>`

### How to test

#### Command without specifying case_id:
- [x] On hasta-stage, run `$ cg delete observations`. This should raise the following error and fail `Error: Missing option "-c" / "--case_id".`

#### Dry-run:

- [x] On hasta-stage, run `$ cg delete observations -c <case_id> --dry-run`

#### Non-existing case:

- [x] On hasta-stage, run `$ cg delete observations -c nonexistingcase`. This should abort with the following message `Case nonexistingcase could not be found in StatusDB!`.

#### Aborting existing case:

- [x] On hasta-stage, run `$ cg delete observations -c <case_id>`. 
- [x] When asked: `Do you want to continue? [y/N]:` type `N`
- [x] On statusdb-stage, under the tab `Sample`, check that `Loqusdb Id` column is still present
- [x] On hasta-stage, check that the case still exists on loqusdb `$ loqusdb cases --case-id <case_id>`

#### Existing case and skip confirmation:

- [x] On hasta-stage, run `$ cg delete observations -c <case_id> --yes`. **NOTE: This will delete all observations for that case in LoqusDB**
- [x] On statusdb-stage, under the tab `Sample`, check that `Loqusdb Id` column is empty
- [x] On hasta-stage, check that the case is no longer in loqusdb `$ loqusdb cases --case-id <case_id>`. The output should return `INFO Case CASE_ID does not exist in database`


### Expected test outcome

- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by @fevac 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [x] Document in [Atlas](https://atlas.scilifelab.se/infrastructure/loqusdb/delete_case/) 
- [ ] Deploy this branch on ...
- [ ] Inform to ...

henrikstranneheim commented on cg/apps/loqus.py:
```diff
@@ -71,6 +71,20 @@ def load(
+        raise DeleteCaseError(f"Could not delete case {case_id}")
+
     def get_case(self, case_id: str) -> dict:
         """Find a case in the database by case id"""
```
```suggestion
        """Find a case in the database by case id."""
```

henrikstranneheim commented on cg/cli/base.py:
```diff
@@ -25,7 +25,7 @@
 from .export import export
```
Let's change all these relative imports in this module to absolute (which is best practice)

henrikstranneheim commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,35 @@
+
+@delete_cmd.command()
+@click.option("-c", "--case_id", help="internal case id")
+@click.pass_obj
```
I would feel more comfortable if there also was a dry-run option that would do all except deleting. Removing stuff is hazardous processes. There is global click dry-run constant in: cg/constants/constants.py for import and reuse

henrikstranneheim commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,35 @@
+
+    # Delete case from loqusdb
+    LOG.info("This will delete all variants in LoqusDB for case: %s", case_id)
+    click.confirm("Do you want to continue?", abort=True)
```
It would be nice to also add this:
```
@click.option("-y", "--yes", is_flag=True, help="Skip confirmation")
```
I bet that is repeated across cg. Could you consider making it a global constant, much like the dry-run option. then we do not have to repeat ourselves everywhere (DRY)

henrikstranneheim commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,35 @@
+    status_db: Store = context.status_db
+    loqusdb_api: LoqusdbAPI = context.loqusdb_api
+
+    # Delete case from loqusdb
```
```suggestion
```
Code is clear enough

henrikstranneheim commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,35 @@
+    LOG.info("Delete variants from LoqusDB for case: %s", case_id)
+    loqusdb_api.delete(case_id=case_id)
+
+    # If a case is deleted reset StatusDB loqusdb_id to None
```
```suggestion
```
Code is clear enough and one could always read the doc string. It is hard to remember to update comments when the code change so they should be used with caution and when they convey context not evident in the code

henrikstranneheim commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,35 @@
+    loqusdb_api.delete(case_id=case_id)
+
+    # If a case is deleted reset StatusDB loqusdb_id to None
+    status_db.reset_observations(case_id)
```
```suggestion
    status_db.reset_loqusdb_observation_ids(case_id)
```

henrikstranneheim commented on tests/apps/loqus/conftest.py:
```diff
@@ -176,3 +189,19 @@ def fixture_loqusdb_duplicate_output() -> bytes:
+    """
+    loqusdb stderr for a successful delete.
+    """
+    return LOQUSDB_DELETE_STDERR
```
```suggestion
    return b"""2022-09-22 12:30:07 evacmbp loqusdb.commands.cli[20689] INFO Running loqusdb version 2.6.9
2022-09-22 12:30:07 evacmbp mongo_adapter.client[20689] INFO Connecting to uri:mongodb://None:None@localhost:27017
2022-09-22 12:30:07 evacmbp mongo_adapter.client[20689] INFO Connection established
2022-09-22 12:30:07 evacmbp mongo_adapter.adapter[20689] INFO Use database loqusdb
2022-09-22 12:30:07 evacmbp loqusdb.plugins.mongo.case[20689] INFO Removing case yellowhog from database
2022-09-22 12:30:07 evacmbp loqusdb.utils.delete[20689] INFO deleting variants
2022-09-22 12:30:07 evacmbp loqusdb.utils.delete[20689] INFO Start deleting chromosome 1"""
```
I would skip the extra constant initialization step

henrikstranneheim commented on tests/apps/loqus/conftest.py:
```diff
@@ -176,3 +189,19 @@ def fixture_loqusdb_duplicate_output() -> bytes:
+@pytest.fixture(name="loqusdb_delete_stderr")
+def fixture_loqusdb_delete_stderr() -> bytes:
+    """
+    loqusdb stderr for a successful delete.
```
```suggestion
    Loqusdb STDERR for a successful delete.
```

henrikstranneheim commented on tests/apps/loqus/conftest.py:
```diff
@@ -176,3 +189,19 @@ def fixture_loqusdb_duplicate_output() -> bytes:
+@pytest.fixture(name="loqusdb_delete_nonexisting_stderr")
+def fixture_loqusdb_delete_nonexisting_stderr() -> bytes:
+    """
+    loqusdb delete stderr for non existing case.
```
```suggestion
    Loqusdb delete STDERR for non existing case.
```

henrikstranneheim commented on tests/apps/loqus/conftest.py:
```diff
@@ -176,3 +189,19 @@ def fixture_loqusdb_duplicate_output() -> bytes:
+    """
+    loqusdb delete stderr for non existing case.
+    """
+    return LOQUSDB_DELETE_NONEXISTING_STDERR
```
Same here

henrikstranneheim commented on tests/apps/loqus/test_loqusdbapi.py:
```diff
@@ -142,3 +142,29 @@ def test_repr_string(loqus_config):
     assert repr_string == correct_string
+
+
+def test_delete(case_id: str, loqusdbapi: LoqusdbAPI, loqusdb_delete_stderr: bytes):
```
```suggestion
def test_delete_case(case_id: str, loqusdbapi: LoqusdbAPI, loqusdb_delete_stderr: bytes):
```

henrikstranneheim commented on tests/apps/loqus/test_loqusdbapi.py:
```diff
@@ -142,3 +142,29 @@ def test_repr_string(loqus_config):
+    loqusdbapi.process.stderr = loqusdb_delete_stderr.decode("utf-8")
+    data = loqusdbapi.delete(case_id)
+
+    # THEN
```
```suggestion
    # THEN None should be returned and no errors raised
```

henrikstranneheim commented on tests/apps/loqus/test_loqusdbapi.py:
```diff
@@ -142,3 +142,29 @@ def test_repr_string(loqus_config):
+    assert data is None
+
+
+def test_delete_nonexisting_case(
```
```suggestion
def test_delete_case_non_existing_case(
```

henrikstranneheim commented on tests/apps/loqus/test_loqusdbapi.py:
```diff
@@ -142,3 +142,29 @@ def test_repr_string(loqus_config):
+    # WHEN deleting an existing case from loqusdb
+    loqusdbapi.process.stderr = loqusdb_delete_nonexisting_stderr.decode("utf-8")
+
+    # THEN assert that the error is raised
```
```suggestion
    # THEN assert that an error is raised
```

henrikstranneheim reviewed (commented):
Very nice! 👍

henrikstranneheim commented:
@fevac It would also be great if you could document this new improved feature in Atlas here: https://atlas.scilifelab.se/infrastructure/loqusdb/delete_case/. We should discourage from interacting with Loqusdb directly since it will cause the bug you are fixing.

ivadym commented on cg/cli/delete/observations.py:
```diff
@@ -0,0 +1,42 @@
+    pass
+
+
+@delete_cmd.command()
```
Yes, if the name of the function is the same we don't need it :)

henrikstranneheim reviewed (approved):
Great work!

Just a functional verification on stage and then let's M&D

fevac commented:
### Upload case `fitraptor` 

```shell
$ cg upload observations --case_id fitraptor
```
<img width="342" alt="image" src="https://user-images.githubusercontent.com/29628428/192818095-09d763e1-f9ce-4a8c-bef8-6aae99922fa6.png">

- [x] Make sure that loqusdb has variables:

```shell
loqusdb cases --case-id fitraptor
```
<img width="1061" alt="image" src="https://user-images.githubusercontent.com/29628428/192819033-8a24d963-3ceb-4a2e-8ed1-def20a3eb5d2.png">

- [x] Make sure Loqusdb Id is set on statusdb stage under the Sample tab. 

<img width="208" alt="image" src="https://user-images.githubusercontent.com/29628428/192818192-8c42e861-cb18-43e5-9333-d10f0aaf5c58.png">

### Missing --case_id

`$ cg delete observations`

<img width="721" alt="image" src="https://user-images.githubusercontent.com/29628428/192819961-f0738d8b-43a7-4f18-884d-0d52904d0231.png">

### Dry-run

`$ cg delete observations -c fitraptor --dry-run`

<img width="803" alt="image" src="https://user-images.githubusercontent.com/29628428/192820133-4527a3c2-21d4-493f-bcc1-1ea7000ac73a.png">

### Non-existing case

`$ cg delete observations -c nonexistingcase`

<img width="1483" alt="image" src="https://user-images.githubusercontent.com/29628428/192822154-4ac96258-a836-49c1-86c4-4510a6255f0f.png">


### Delete case 

```shell
cg delete observations -c fitraptor
```

fevac commented:
- I changed the behaviour when a case doesn't exists. Instead of raising an error, it now prints a message and aborts.
- I took KN suggestion and added a some checks to make sure the case exists in both statusdb and loqusdb prior attempting to delete.

fevac commented:
Atlas updated on PR: https://github.com/Clinical-Genomics/atlas/pull/1106

fevac commented:
Deployed in production

<img width="593" alt="image" src="https://user-images.githubusercontent.com/29628428/193787822-0a820e31-598b-4bee-97de-d41f4ce32bbe.png">
